### DISCUSSION TITLE
How to ensure the consistency of substructure bit in fingerprints of different molecules

### DISCUSSION THREAD

**Discussion by zwang1995** [rdkit/rdkit#4688] [Q&A] [answered]
Two molecules have different substructures in bit-1 (see image attached below).

If I have a large number of molecules, how to force them to have the same substructure in the same bit? I want to use them for Machine Learning, so I have to ensure that all these fingerprint vectors share the same substructure information.

```
mol1 = Chem.MolFromSmiles(Chem.CanonSmiles("OCCC"))

bitinfo1 = {}
fp = AllChem.GetMorganFingerprintAsBitVect(mol1, 1, nBits=3, bitInfo=bitinfo1)

fp_arr = np.zeros((1,))
DataStructs.ConvertToNumpyArray(fp, fp_arr)
print(fp_arr, bitinfo)
tpls = [(mol1 ,x, bitinfo1) for x in fp.GetOnBits()]
Draw.DrawMorganBits(tpls, molsPerRow=4, legends=[str(x) for x in fp.GetOnBits()])
```

```
mol2 = Chem.MolFromSmiles(Chem.CanonSmiles("CCC"))

bitinfo2 = {}
fp = AllChem.GetMorganFingerprintAsBitVect(mol2, 1, nBits=3, bitInfo=bitinfo2)

fp_arr = np.zeros((1,))
DataStructs.ConvertToNumpyArray(fp, fp_arr)
print(fp_arr, bitinfo2)
tpls = [(mol2 ,x, bitinfo2) for x in fp.GetOnBits()]
Draw.DrawMorganBits(tpls, molsPerRow=4, legends=[str(x) for x in fp.GetOnBits()])
```

![image](https://user-images.githubusercontent.com/55940649/141284044-2029b875-bb4b-46a5-8596-1bf489db2fc4.png)

**Comment by greglandrum** [comment, marked as answer]
By reducing the fingerprint to be only three bits long you've guaranteed that every possible atom environment will be mapped into only three possible positions. That's almost certainly not what you want.

The Morgan fingerprint algorithm will guarantee that the same atom environment always maps to the same bit. What it does not guarantee is that a single bit corresponds only to a single substructure. There's a lot of information about this topic in this mailing-list thread:
https://sourceforge.net/p/rdkit/mailman/message/36438523/

**Reply by zwang1995** [reply]
Mant thanks!
